### PR TITLE
Refactor SuggestionsList to use hooks and change class to function component

### PR DESCRIPTION
**Pull request by amustaque97**
## Description

Introduced functional component. In the component I'm using useState, useEffect.
Renamed a few variables to get rid of syntax errors/conflicts. Tested the component and
it is working as expected.

## How has this been tested?

1. In wp admin go to tags page.
image
2. Create multiple tags
image

3. Edit or create new post.

4. In the settings, try to add new tags.
5. It will show  suggestion list drop-down
image

6. Select few tags 
7. Publish the post.
## Types of changes

Code Quality / Refactor
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/form-token-field/test/index.js [hunk 1]
```diff
@@ -262,7 +262,9 @@ describe( 'FormTokenField', () => {
 
 			// before sending a hover event, we need to wait for
 			// SuggestionList#_scrollingIntoView to become false
-			jest.advanceTimersByTime( 100 );
+			act( () => {
+				jest.advanceTimersByTime( 100 );
+			} );
```
Comment (ciampo) [inline packages/components/src/form-token-field/test/index.js:267]: Why was this change necessary?

**Review by ciampo — COMMENTED** [review]
Thank you @amustaque97 for working on this ✌️ 

I'm going to give this a proper look later (either by end of day or on Friday), but I gave a quick look and code changes overall look good to me.

File: packages/components/src/form-token-field/test/index.js (hunk 1 above)
Comment (amustaque97) [reply packages/components/src/form-token-field/test/index.js:267]: Half of the month is gone. Don't remember exact details 😅 . Sorry about that. But yeah test case was failing and after going to test failure. I learned more about act and I was able to fix it. I come to know about `act` from test failure error message as shown in the below image.
Screenshot 2021-12-16 at 12 24 17 AM

**Comment by ciampo** [comment]
Hey @amustaque97, sorry if I haven't managed to take a proper look at this yet, but I had a few unforeseen PRs that took precedence. I will probably be able to do to in the new year, since I'm going to be AFK in the next couple of weeks. Hope that's ok!

File: packages/components/src/form-token-field/suggestions-list.js [hunk 2]
```diff
@@ -2,157 +2,147 @@
...
-		super( ...arguments );
-		this.handleMouseDown = this.handleMouseDown.bind( this );
-		this.bindList = this.bindList.bind( this );
-	}
+function SuggestionsList( {
+	selectedIndex,
+	scrollIntoView,
+	match = '',
+	onHover = () => {},
+	onSelect = () => {},
```
Comment (ZebulanStanphill) [inline packages/components/src/form-token-field/suggestions-list.js]: A slight optimization could be made by using lodash's `noop` here to reduce the number of functions being created, e.g.:

```js
import { map, noop } from 'lodash'
// ...
	onHover = noop,
	onSelect = noop,
```

File: packages/components/src/form-token-field/suggestions-list.js [hunk 3]
```diff
@@ -2,157 +2,147 @@
...
+	suggestions = Object.freeze( [] ),
```
Comment (ZebulanStanphill) [inline packages/components/src/form-token-field/suggestions-list.js]: `Object.freeze( [] )` could be moved out of this function into a `const` to avoid recreating it on every render.

File: packages/components/src/form-token-field/suggestions-list.js [hunk 4]
```diff
@@ -2,157 +2,147 @@
...
+	const handleClick = ( suggestion ) => {
 		return () => {
-			this.props.onSelect( suggestion );
+			onSelect( suggestion );
 		};
-	}
+	};
 
-	handleMouseDown( e ) {
+	const handleMouseDown = ( e ) => {
```
Comment (ZebulanStanphill) [inline packages/components/src/form-token-field/suggestions-list.js]: This function could be moved to the top level of the file to avoid recreating it on every render.

**Comment by amustaque97** [comment]
All review comments looks logical to me. I have addressed review comments. Thank You for the review. :smile:

File: packages/components/src/form-token-field/suggestions-list.js [hunk 5]
```diff
@@ -1,158 +1,150 @@
...
+	scrollIntoView,
+	match = '',
+	onHover = noop,
+	onSelect = noop,
+	suggestions = emptyList,
+	displayTransform,
+	instanceId,
+	setTimeout,
+} ) {
+	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
```
Comment (ciampo) [inline packages/components/src/form-token-field/suggestions-list.js]: Shouldn't the default value of `scrollingIntoView` be `false` ?

File: packages/components/src/form-token-field/suggestions-list.js [hunk 6]
```diff
@@ -1,158 +1,150 @@
...
+	const [ list, setList ] = useState();
```
Comment (ciampo) [inline packages/components/src/form-token-field/suggestions-list.js]: Instead of using state and a custom ref callback to store a `ref` to the `<ul>` element, we can use the [custom `useRefEffect` hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts):

<details>
<summary>Click to expand suggested changes</summary>

```diff
diff --git a/packages/components/src/form-token-field/suggestions-list.js b/packages/components/src/form-token-field/suggestions-list.js
index 397a33863e..1c0d770af6 100644
--- a/packages/components/src/form-token-field/suggestions-list.js
+++ b/packages/components/src/form-token-field/suggestions-list.js
@@ -8,8 +8,8 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { useEffect, useState } from '@wordpress/element';
-import { withSafeTimeout } from '@wordpress/compose';
+import { useState } from '@wordpress/element';
+import { withSafeTimeout, useRefEffect } from '@wordpress/compose';
 
 const emptyList = Object.freeze( [] );
 
@@ -30,30 +30,27 @@ function SuggestionsList( {
 	setTimeout,
 } ) {
 	const [ scrollingIntoView, setScrollingIntoView ] = useState( true );
-	const [ list, setList ] = useState();
+	const listRef = useRefEffect(
+		( listNode ) => {
+			// only have to worry about scrolling selected suggestion into view
+			// when already expanded
+			if (
+				selectedIndex > -1 &&
+				scrollIntoView &&
+				listNode.children[ selectedIndex ]
+			) {
+				setScrollingIntoView( true );
+				scrollView( listNode.children[ selectedIndex ], listNode, {
+					onlyScrollIfNeeded: true,
+				} );
 
-	useEffect( () => {
-		// only have to worry about scrolling selected suggestion into view
-		// when already expanded
-		if (
-			selectedIndex > -1 &&
-			scrollIntoView &&
-			list.children[ selectedIndex ]
-		) {
-			setScrollingIntoView( true );
-			scrollView( list.children[ selectedIndex ], list, {
-				onlyScrollIfNeeded: true,
-			} );
-
-			setTimeout( () => {
-				setScrollingIntoView( false );
-			}, 100 );
-		}
-	}, [ selectedIndex, scrollIntoView, list ] );
-
-	const bindList = ( ref ) => {
-		setList( ref );
-	};
+				setTimeout( () => {
+					setScrollingIntoView( false );
+				}, 100 );
+			}
+		},
+		[ selectedIndex, scrollIntoView ]
+	);
 
 	const handleHover = ( suggestion ) => {
 		return () => {
@@ -98,7 +95,7 @@ function SuggestionsList( {
 	// TODO does this still apply now that it's a <ul> and not a <div>?
 	return (
 		<ul
-			ref={ bindList }
+			ref={ listRef }
 			className="components-form-token-field__suggestions-list"
 			id={ `components-form-token-suggestions-${ instanceId }` }
 			role="listbox"
```
</details>

This could potentially help to fix the [e2e test error](https://github.com/WordPress/gutenberg/runs/4575862072?check_suite_focus=true), since the callback in `useRefEffect` is invoked only if `listNode` isn't falsy.

File: packages/components/src/form-token-field/suggestions-list.js [hunk 7]
```diff
@@ -1,158 +1,150 @@
...
+	return (
+		<ul
+			ref={ bindList }
+			className="components-form-token-field__suggestions-list"
+			id={ `components-form-token-suggestions-${ instanceId }` }
+			role="listbox"
+		>
+			{ map( suggestions, ( suggestion, index ) => {
+				const matchText = computeSuggestionMatch( suggestion );
+				const classeName = classnames(
```
Comment (ciampo) [inline packages/components/src/form-token-field/suggestions-list.js]: Nit: typo

```suggestion
				const className = classnames(
```

(we should then update any reference to this variable)

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, @amustaque97 !

I took some time, had a proper look at the code and left some comments

File: packages/components/src/form-token-field/suggestions-list.js (hunk 2 above)
Comment (ciampo) [reply packages/components/src/form-token-field/suggestions-list.js]: Premise: nit

Just wanted to add that another approach is to leave those function props as `undefined` by default, and use the optional chaining syntax when invoking them, e.g.

```js
const handHover = () => {
  // do stuff

  onHover?.();
}
```

This has the the advantage of relying solely on native APIs/syntax (plus it avoids initialising those values to their default value). But this is more of a personal preference / micro-optimisation, so we can keep the code as-is.

Comment (ZebulanStanphill) [reply packages/components/src/form-token-field/suggestions-list.js]: I agree that using optional chaining and an `undefined` default is better than using lodash's `noop`. Thanks for pointing that out.

Comment (amustaque97) [reply packages/components/src/form-token-field/suggestions-list.js]: Thanks for the helping me here 😅 I would prefer to keep things simple. Your comment will help others as well to understand of writing this way instead of `noop`

File: packages/components/src/form-token-field/suggestions-list.js (hunk 5 above)
Comment (amustaque97) [reply packages/components/src/form-token-field/suggestions-list.js]: Oops! you're right 💯 

Correct me If I'm wrong here, I checked variable here https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/form-token-field/index.js#L50
and based on that I'm agreeing its value should be false.

File: packages/components/src/form-token-field/suggestions-list.js (hunk 7 above)
Comment (amustaque97) [reply packages/components/src/form-token-field/suggestions-list.js]: Apologies for this 😅 

I have tested again my changes. Working as expected. :smile:

File: packages/components/src/form-token-field/suggestions-list.js (hunk 6 above)
Comment (amustaque97) [reply packages/components/src/form-token-field/suggestions-list.js]: Can you please help me in understanding where it was wrong in my code? Just curious

File: packages/components/src/form-token-field/suggestions-list.js (hunk 5 above)
Comment (ciampo) [reply packages/components/src/form-token-field/suggestions-list.js]: My reasoning was that, prior to your refactor, `this.scrollingIntoView` was not initialised to anything — and therefore it held a value of `undefined`. Since we're refactoring the code to `useState` and we want that variable to always hold a boolean value, then we should use `false` (since `undefined` is a falsy value)

Hope this is clear enough!

File: packages/components/src/form-token-field/suggestions-list.js (hunk 6 above)
Comment (ciampo) [reply packages/components/src/form-token-field/suggestions-list.js]: There wasn't anything _wrong_ in your code — you were just refactoring the component from class to function — my suggestion here is more of an improvement on top of your changes. I believe the code is cleaner (and potentially causes less re-renders)

Here's the reasoning that brought me to suggest `useRefEffect`:

- While reviewing your PR, I was having a look at the `list` state, and noticed that it was used to hold a `ref`.
- Now, we could have used `useRef` instead, but then I noticed that `list` was used as a dependency of the `useEffect` hook — switching to `useRef` would meant that changes to `list` wouldn't trigger the `useEffect` hook.
- The proper way to refactor the code, then, would be to rewrite `list` to use `useCallback` to hold the `ref`. The code in `useEffect` (and its dependencies) could have moved to the body of `useCallback` too
- Finally, I realised that the pattern that I just described is implemented in a custom hook called [`useRefEffect`](https://github.com/WordPress/gutenberg/blob/trunk/packages/compose/src/hooks/use-ref-effect/index.ts)

File: packages/components/src/form-token-field/suggestions-list.js (hunk 2 above)
Comment (ciampo) [reply packages/components/src/form-token-field/suggestions-list.js]: Just seen your last commit — I believe that the alternative to using `noop`  that we agreed on should look more like this:

```diff
diff --git a/packages/components/src/form-token-field/suggestions-list.js b/packages/components/src/form-token-field/suggestions-list.js
index 3794448014..b7eee38e95 100644
--- a/packages/components/src/form-token-field/suggestions-list.js
+++ b/packages/components/src/form-token-field/suggestions-list.js
@@ -22,8 +22,8 @@ function SuggestionsList( {
 	selectedIndex,
 	scrollIntoView,
 	match = '',
-	onHover = () => {},
-	onSelect = () => {},
+	onHover,
+	onSelect,
 	suggestions = emptyList,
 	displayTransform,
 	instanceId,
@@ -55,14 +55,14 @@ function SuggestionsList( {
 	const handleHover = ( suggestion ) => {
 		return () => {
 			if ( ! scrollingIntoView ) {
-				onHover( suggestion );
+				onHover?.( suggestion );
 			}
 		};
 	};
 
 	const handleClick = ( suggestion ) => {
 		return () => {
-			onSelect( suggestion );
+			onSelect?.( suggestion );
 		};
 	};
 
```

File: packages/components/src/form-token-field/suggestions-list.js (hunk 6 above)
Comment (amustaque97) [reply packages/components/src/form-token-field/suggestions-list.js]: Thanks for the brief explanation.

**Review by ciampo — COMMENTED** [review]
Thank you for addressing the feedback! Not crucial, but in the future it would be even better if you could split different sets of changes over separate commits

I just left a minor comment, otherwise the code changes LGTM and tests well.

Could you also add an entry to the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) ?

**Review by ciampo — APPROVED** [review]
